### PR TITLE
Add counter for errors and change the entry point of exporter metrics

### PR DESCRIPTION
**Pull request by daviddetorres**
This PR implements #4 and its comments on:
- Add counter for the different possible errors (connection, timeout...)
- Moving the exporter metrics from :0911/metrics to :9602/metrics
- Moving the target metrics from :9602/metrics to :9602/modbus?target=1.2.3.4&module=module_name&sub_target=subtarget_number

Documentation of the new metrics of the exporter:
- Added counter modbus_requests_total with labels for:
   - target
   - request status

- Added counter modbus_request_duration_seconds_total with labels for:
    - target

- Added possible requests status:
   - OK:  successful
   - ERROR_SOCKET: error opening socket connection
   - ERROR_TIMEOUT: connection established but no response from modbus device
   - ERROR_PARSING_VALUE: error parsing value received

Error status start with "ERROR" so it is easy to filter in promQL.

### CODE REVIEW

File: modbus_exporter.go [hunk 1]
```diff
@@ -31,10 +33,29 @@ import (
...
+	ModbusRequestStatusOK ModbusRequestStatusType = "OK"
+	// ModbusRequestStatusErrorSock error opening socket connection
+	ModbusRequestStatusErrorSock ModbusRequestStatusType = "ERROR_SOCKET"
+	// ModbusRequestStatusErrorTimeout connection established but no response from modbus device
+	ModbusRequestStatusErrorTimeout ModbusRequestStatusType = "ERROR_TIMEOUT"
+	// ModbusRequestStatusErrorParsingValue error parsing value received
+	ModbusRequestStatusErrorParsingValue ModbusRequestStatusType = "ERROR_PARSING_VALUE"
+)
+
+var (
```
Comment (mxinden) [inline modbus_exporter.go]: I would try to avoid global metric variables.

Comment (RichiH) [reply modbus_exporter.go]: Agreed.

Comment (RichiH) [reply modbus_exporter.go:50]: Any update here?

Comment (RichiH) [reply modbus_exporter.go:50]: @daviddetorres

File: modbus_exporter.go [hunk 2]
```diff
@@ -45,20 +66,28 @@ func main() {
...
+		Name: "modbus_request_duration_seconds_total",
+		Help: "Total duration of modbus successful requests by target in seconds",
+	}, []string{"target"})
+	telemetryRegistry.MustRegister(modbusDurationCounterVec)
+
+	modbusRequestsCounterVec = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "modbus_requests_total",
+		Help: "Number of modbus request by status and target",
+	}, []string{"target", "status"})
+	telemetryRegistry.MustRegister(modbusRequestsCounterVec)
```
Comment (SuperQ) [inline modbus_exporter.go:79]: This shouldn't be necessary, as these metrics are covered by the `scrape_duration_seconds` and `up` metric provided by Prometheus itself.

Comment (bastischubert) [reply modbus_exporter.go:79]: @superq  What about other monitoring systems that implement openmetrics ,don't have that as internal metrics and just want to scrape the exporter?

**Comment by RichiH** [comment]
Merging as a basis for @SuperQ to PR from. Thanks @daviddetorres !

**Comment by rfc1036** [comment]
The test `TestScrapeHandler` fails after merging this commit.

```
go build
./modbus_exporter --help 2> help.txt || true
/home/md/go/bin/embedmd -w README.md
rm help.txt
go test ./...
time="2020-12-08T07:42:27+01:00" level=info msg="got scrape request for module 'my_module' target '10.0.0.10' and sub_target '10'" source="modbus_exporter.go:130"
--- FAIL: TestScrapeHandler (0.00s)
    --- FAIL: TestScrapeHandler/module_and_target (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7955a6]

goroutine 12 [running]:
testing.tRunner.func1.1(0x87caa0, 0xc0ad20)
	/usr/lib/go-1.15/src/testing/testing.go:1072 +0x30d
testing.tRunner.func1(0xc000192480)
	/usr/lib/go-1.15/src/testing/testing.go:1075 +0x41a
panic(0x87caa0, 0xc0ad20)
	/usr/lib/go-1.15/src/runtime/panic.go:969 +0x1b9
github.com/prometheus/client_golang/prometheus.(*CounterVec).GetMetricWithLabelValues(...)
	/home/md/go/pkg/mod/github.com/prometheus/client_golang@v0.9.2/prometheus/counter.go:172
github.com/prometheus/client_golang/prometheus.(*CounterVec).WithLabelValues(0x0, 0xc00000f820, 0x2, 0x2, 0x0, 0xc00002c800)
	/home/md/go/pkg/mod/github.com/prometheus/client_golang@v0.9.2/prometheus/counter.go:204 +0x26
github.com/lupoDharkael/modbus_exporter.scrapeHandler(0xc0001a1ec8, 0x986580, 0xc00006d0c0, 0xc000178400)
	/USR3/src/modbus/modbus_exporter/modbus_exporter.go:137 +0x872
github.com/lupoDharkael/modbus_exporter.TestScrapeHandler.func5(0xc000192480)
	/USR3/src/modbus/modbus_exporter/modbus_exporter_test.go:97 +0x3cd
testing.tRunner(0xc000192480, 0xc00018e750)
	/usr/lib/go-1.15/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/usr/lib/go-1.15/src/testing/testing.go:1168 +0x2b3
FAIL	github.com/lupoDharkael/modbus_exporter	0.006s
ok  	github.com/lupoDharkael/modbus_exporter/config	(cached)
--- FAIL: TestParseModbusData (0.00s)
    --- FAIL: TestParseModbusData/bool,_first_bit (0.00s)
        modbus_test.go:465: expected metric value to be 1 but got 0
    --- FAIL: TestParseModbusData/bool,_second_bit (0.00s)
        modbus_test.go:465: expected metric value to be 1 but got 0
FAIL
FAIL	github.com/lupoDharkael/modbus_exporter/modbus	0.004s
?   	github.com/lupoDharkael/modbus_exporter/tests/fake_server	[no test files]
FAIL
make: *** [Makefile:33: test] Errore 1
```

File: modbus_exporter.go (hunk 2 above)
Comment (SuperQ) [reply modbus_exporter.go:79]: That's up to them to implement. This is duplicate information.

**Pull request merged**